### ISSUE TITLE
Metric calculation node development

### ISSUE THREAD

**Issue by mitsudome-r** [autowarefoundation/autoware-projects#11]
## Description
(Copied issue from https://gitlab.com/autowarefoundation/autoware.auto/AutowareAuto/-/issues/1446)
To implement and upload to AutowreAuto repository sample ROS2 node for metric calculation. As a sample, the calculation of deviation from planned path should be calculated.

### Purpose
To validate specific functionalities of AutowareAuto

### Desired Behavior
Metric calculation node subscribes to ROS2  topics published by Autoware or Simulation pipeline and publishes value of calculated metric

### Definition of Done  
- [ ] Sample metric calculation node developed and uploaded to the AutowareAuto repository
- [ ] Instruction how to run, and how to develop other nodes for calculation of different metrics

**Comment by WJaworskiRobotec** [comment]
Similar feature related to trajectory evaluation : https://github.com/autowarefoundation/autoware.universe/pull/288

**Comment by domrobotec** [comment]
Currently working on this issue. 
I am reusing parts of the planning_evaluator and motion_evaluator and expanding it to calculate various metrics, like:
* average velocity
* calculation of localization errors: lateral, absolute deviations from the trajectory
* heading angle error
* distance to closest obstacle

The final goals are:
* to have a generic metric calculator that more specific metric calculators (planning, localization, object detection, possibly more?) could inherit from.
* to make it easy to add new metrics under different categories (planning, localization, object detection, etc.)

**Comment by domrobotec** [comment]
Recently added 2 separate evaluators (localization and kinematic), cleaned the code, and added basic unit tests.

I am currently trying to find the best way to find proper beginning and ending times between which metrics should be collected as it should happen only when a vehicle is driving. 
Additionally, working on segmentation evaluator and extracting common parts of all evaluators to parent evaluator.

**Comment by domrobotec** [comment]
Added draft of ground detection error calculation from point clouds - calculating confusion matrix.
Also, tested localization evaluator with TierIV SIM.

**Comment by WJaworskiRobotec** [comment]
@djargot Have you added publishing of gt localization on the seperate topic (/tf_ground_truth or sth like that) to a branch in TierIV simulator ?

**Comment by WJaworskiRobotec** [comment]
Localization and kinematic evaluators are already merged to autoware.universe

[localization and kinematic evaluator video](https://drive.google.com/file/d/1Q45RTPO11Jbul-28vKeDfMFw1gCIl8xl/view?usp=sharing)

**Comment by domrobotec** [comment]
I created initial version of segmentation metrics (https://github.com/autowarefoundation/autoware.universe/pull/1600) to calculate precision, recall, accuracy, and confusion matrix. Values are calculated both for a single point cloud as well as across whole scenario/ROS bag. 

I tested metrics using labeled point cloud provided by MORAI (related issue: https://github.com/autowarefoundation/autoware.universe/issues/669).

Example results for ground segmentation metrics for a single point cloud:
![confmat](https://user-images.githubusercontent.com/103112671/185134875-016b07fd-0f7b-459d-98d7-6f7bc1c317e1.png)
